code/datums/components/orbiter.dm

```diff
@@ -0,0 +1,237 @@
+
+	//Head first!
+	if(pre_rotation)
+		var/matrix/M = matrix(orbiter.transform)
+		var/pre_rot = 90
+		if(!clockwise)
+			pre_rot = -90
+		M.Turn(pre_rot)
+		orbiter.transform = M
+
+	var/matrix/shift = matrix(orbiter.transform)
+	shift.Translate(0,radius)
+	orbiter.transform = shift
+
+	SEND_SIGNAL(parent, COMSIG_ATOM_ORBIT_BEGIN, orbiter)
+
+	// If we changed orbits, we didn't stop our rotation, and don't need to start it up again
+	if(!was_refreshing)
+		orbiter.SpinAnimation(rotation_speed, -1, clockwise, rotation_segments, parallel = FALSE)
+
+	var/target_loc = get_turf(parent)
+	if(force_move)
+		orbiter.forceMove(target_loc)
+	else
+		orbiter.loc = target_loc
+
+/**
+* End the orbit and clean up our transformation.
+* If this removes the last atom orbiting us, then qdel ourselves.
+* If refreshing == TRUE, variables will be cleaned up as necessary, but src won't be qdeled.
+*/
+/datum/component/orbiter/proc/end_orbit(atom/movable/orbiter, refreshing = FALSE)
+	if(!(orbiter in orbiter_list))
+		return
+
+	var/matrix/cached_transform = transform_cache[orbiter]
+
+	orbiter.transform = cached_transform
+
+	orbiter.orbiting = null
+	orbiter.orbit_params = 0
+	orbiter.stop_orbit()
+	UnregisterSignal(orbiter, COMSIG_MOVABLE_MOVED)
+
+	SEND_SIGNAL(parent, COMSIG_ATOM_ORBIT_STOP, orbiter)
+
+	orbiter_list -= orbiter
+	transform_cache -= orbiter
+
+	if(!length(orbiter_list) && !QDELING(src) && !refreshing)
+		qdel(src)
+
+/// Called when the orbited user moves
+/datum/component/orbiter/proc/parent_move_react(atom/movable/orbited, atom/old_loc, direction)
+
+	set waitfor = FALSE // Transfer calls this directly and it doesnt care if the ghosts arent done moving
+
+	if(orbited.loc == old_loc)
+		return
+
+	var/turf/new_turf = get_turf(orbited)
+	if(!new_turf)
+		qdel(src)
+
+	var/atom/cur_loc = orbited.loc
+	for(var/atom/movable/movable_orbiter in orbiter_list)
+		if(QDELETED(movable_orbiter) || movable_orbiter.loc == new_turf)
+			continue
+
+		if(movable_orbiter.orbit_params & ORBIT_FORCE_MOVE)
+			movable_orbiter.forceMove(cur_loc)
+		else
+			movable_orbiter.loc = cur_loc
+
+		if(CHECK_TICK && orbited.loc != cur_loc)
+			// We moved again during the checktick, cancel current operation
+			break
+
+/**
+* Called when the orbiter themselves moves
+*/
+/datum/component/orbiter/proc/orbiter_move_react(atom/movable/orbiter, atom/oldloc, direction)
+	SIGNAL_HANDLER
+
+	if(orbiter.loc == get_turf(parent))
+		return
+
+	if(orbiter.orbiting && orbiter.orbiting == src)
+		// Only end the spin animation when we're actually ending an orbit, not just changing targets
+		orbiter.SpinAnimation(0, 0, parallel = FALSE)
+		end_orbit(orbiter)
+
+
+/////////////////////////////////////////
+// Atom procs/vars
+
+/// Who the current atom is orbiting
+/atom/movable/var/datum/component/orbiter/orbiting = null
+/atom/movable/var/orbit_params = 0
+
+/// who's orbiting the current atom
+/atom/var/datum/component/orbiter/orbiters = null
```

Comment:
I really don't like having 3 vars on atoms/movables for this component. Atoms and components should *ideally* have no references to each other, and only interact via signals. I would love to see it refactored that way if possible. 

At the very least, the `orbit_params` var can probably move off of `atom/movable` and onto the component itself because it's only used in this file.

Reply:
That's fair! I think the old-style used to drop a few variables onto the atoms (and the reference implementation does this as well) but I agree it'd probably be nicer to have it clean up like that.

Reply:
I've been giving it a go, and while it seems to mostly work out nicely, I can't seem to completely get rid of references back to the atom currently being orbited, since that provides necessary context when an orbit is ended (how does the user know which atom to stop orbiting without scanning all other atoms?). I'm going to probably have to leave one variable attached for this case, but should be able to clean the others up.

Reply:
So I've removed most direct component references, with the only new variable being stored on an atom is a UID of the atom being followed. It should only be locate()'d when we need to stop an orbit, and its presence alone should help us know if we're orbiting something.